sql/core/src/test/scala/org/apache/spark/sql/sources/BucketedReadSuite.scala

```diff
@@ -1012,4 +1014,43 @@ abstract class BucketedReadSuite extends QueryTest with SQLTestUtils {
       }
     }
   }
+
+  test("SPARK-32859: disable unnecessary bucketed table scan based on query plan") {
+    withTable("t1", "t2") {
+      df1.write.format("parquet").bucketBy(8, "i").saveAsTable("t1")
+      df2.write.format("parquet").bucketBy(4, "i").saveAsTable("t2")
+
+      def checkNumBucketedScan(query: String, expectedNumBucketedScan: Int): Unit = {
+        val plan = sql(query).queryExecution.executedPlan
+        val bucketedScan = plan.collect { case s: FileSourceScanExec if s.bucketedScan => s }
+        assert(bucketedScan.length == expectedNumBucketedScan)
+      }
+
+      Seq(
+        ("SELECT * FROM t1 JOIN t2 ON t1.i = t2.i", 1, 2),
+        ("SELECT * FROM t1 JOIN t2 ON t1.i = t2.j", 1, 2),
+        ("SELECT * FROM t1 JOIN t2 ON t1.j = t2.j", 0, 2),
+        ("SELECT SUM(i) FROM t1 GROUP BY i", 1, 1),
+        ("SELECT SUM(i) FROM t1 GROUP BY j", 0, 1),
+        ("SELECT * FROM t1 WHERE i = 1", 1, 1),
+        ("SELECT * FROM t1 WHERE j = 1", 0, 1),
```

Comment:
I left two comments about the test;
 - Could you add more test cases, e.g., multiple join cases, multiple bucket column cases, ...?
 - Could you split this single test unit into multiple ones having meaningful test titles?, e.g., `test("SPARK-32859: disable unnecessary bucketed table scan based on query plan - multiple join test")`

Reply:
@maropu - sure. Sorry about missing the comment somehow, not shown up in github page. Added and split tests into 4 (basic, multiple join, multiple bucketed columns, other operators/cases). Thanks.